phpunit/class-wp-get-block-css-selectors-test.php

```diff
@@ -0,0 +1,319 @@
+		$block_type = self::register_test_block(
+			'test/block-with-selectors',
+			array( 'root' => '.wp-custom-block-class' )
+		);
+
+		$selector = wp_get_block_css_selector( $block_type );
+		$this->assertEquals( '.wp-custom-block-class', $selector );
+	}
+
+	public function test_get_root_selector_via_experimental_property() {
+		$block_type = self::register_test_block(
+			'test/block-without-selectors',
+			null,
+			array( '__experimentalSelector' => '.experimental-selector' )
+		);
+
+		$selector = wp_get_block_css_selector( $block_type );
+		$this->assertEquals( '.experimental-selector', $selector );
+	}
+
+	public function test_default_root_selector_generation() {
+		$block_type = self::register_test_block(
+			'test/without-selectors-or-supports',
+			null,
+			null
+		);
+
+		$selector = wp_get_block_css_selector( $block_type );
+		$this->assertEquals( '.wp-block-test-without-selectors-or-supports', $selector );
+	}
+
+	public function test_get_duotone_selector_via_selectors_api() {
+		$block_type = self::register_test_block(
+			'test/duotone-selector',
+			array(
+				'color' => array( 'duotone' => '.duotone-selector' ),
+			),
+			null
+		);
+
+		$selector = wp_get_block_css_selector( $block_type, 'duotone' );
+		$this->assertEquals( '.duotone-selector', $selector );
+	}
+
+	public function test_get_duotone_selector_via_experimental_property() {
+		$block_type = self::register_test_block(
+			'test/experimental-duotone-selector',
+			null,
+			array(
+				'color' => array(
+					'__experimentalDuotone' => '.experimental-duotone',
+				),
+			)
+		);
+
+		$selector = wp_get_block_css_selector( $block_type, 'duotone' );
+		$this->assertEquals( '.experimental-duotone', $selector );
+	}
+
+	public function test_no_duotone_selector_set() {
+		$block_type = self::register_test_block(
+			'test/null-duotone-selector',
+			null,
+			null
+		);
+
+		$selector = wp_get_block_css_selector( $block_type, 'duotone' );
+		$this->assertEquals( '', $selector );
+	}
+
+	public function test_get_feature_selector_via_selectors_api() {
+		$block_type = self::register_test_block(
+			'test/feature-selector',
+			array(
+				'root'       => '.root',
+				'typography' => array( 'root' => '.typography' ),
+			),
+			null
+		);
+
+		$selector = wp_get_block_css_selector( $block_type, 'typography' );
+		$this->assertEquals( '.typography', $selector );
+	}
+
+	public function test_get_feature_selector_via_selectors_api_shorthand_property() {
+		$block_type = self::register_test_block(
+			'test/shorthand-feature-selector',
+			array(
+				'root'       => '.root',
+				'typography' => '.typography',
+			),
+			null
+		);
+
+		$selector = wp_get_block_css_selector( $block_type, 'typography' );
+		$this->assertEquals( '.typography', $selector );
+	}
+
+	public function test_no_feature_level_selector_via_selectors_api() {
+		$block_type = self::register_test_block(
+			'test/null-feature-selector',
+			array( 'root' => '.fallback-root-selector' ),
```

Comment:
It'd be good to also cover the case in which there's no `root` selector so the block fallbacks to the one generated based on the name.

Reply:
I felt like this was covered in the test for the fallback when the block type doesn't use the selectors API. That said, I've added an extra test to be sure of our coverage (https://github.com/WordPress/gutenberg/pull/46496/commits/829658a8336eba0c7524d072498b9c6a6f6dcd22), thanks.